# CM4125 Topic 8 Lab - Running Maps (Solved)

In [ ]:
# Importing the necessary modules
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

## Mapbox

To use maps in Plotly and generate running maps, you need a **Mapbox** account!

You will need to:

- Create a [mapbox.com](https://www.mapbox.com/) account
- Copy your public token from [https://account.mapbox.com/](https://account.mapbox.com/)
- Paste the token into a new file called `mapbox.txt` in the same directory as your notebooks (or upload it to the colab notebook using the cell below)

In [ ]:
## If using google colab, use this code to upload the token
from google.colab import files
uploaded = files.upload()
px.set_mapbox_access_token(open("/content/mapbox.txt").read())

Saving mapbox.txt to mapbox.txt


In [ ]:
## If running locally, only use the following to import (assume the token is in the same directory as the notebook)
px.set_mapbox_access_token(open("mapbox.txt").read())

## Maps in Plotly Express

We are going to start with a dataset of weather from stations in Canada

Compared to the one seen in W3, this one has latitude and longitude data!

In [ ]:
canada = pd.read_csv('https://www.dropbox.com/s/edbw6j60vgcwr3c/canada-weather-2.csv?raw=1')
canada

Station Name  Latitude  Longitude Province  \
0                   ALBERT HEAD    48.400   -123.483       BC   
1        BAMBERTON OCEAN CEMENT    48.583   -123.517       BC   
2                    BEAR CREEK    48.500   -124.000       BC   
3      CENTRAL SAANICH VEYANESS    48.583   -123.417       BC   
4      CENTRAL SAANICH ISL VIEW    48.571   -123.373       BC   
...                         ...       ...        ...      ...   
95090                   GOOSE A    53.317    -60.417       NL   
95091            HOPEDALE (AUT)    55.450    -60.217       NL   
95092          MARY'S HARBOUR A    52.303    -55.848       NL   
95093                      NAIN    56.550    -61.683       NL   
95094                  WABUSH A    52.923    -66.865       NL   

       Mean Temperature (C)  Total Precipitation (mm)  Year  
0                       NaN                       5.0  1971  
1                       NaN                       9.7  1971  
2                      15.4                      20.9  1971  
3                       NaN                       9.4  1971  
4                       NaN                      11.4  1971  
...                     ...                       ...   ...  
95090                  15.8                     109.0  2017  
95091                  11.6                      83.2  2017  
95092                  14.5                      56.9  2017  
95093                  10.6                      38.3  2017  
95094                  12.9                     129.0  2017  

[95095 rows x 7 columns]

Plotly express supports more than one type of map plot

In this lab we will cover `px.scatter_mapbox`, which creates running scatterplots in maps

Here are some links to documentation for two more map types:

- [px.choropleth](http://campusmoodle.rgu.ac.uk/mod/url/view.php?id=3529826)
- [px.line_geo](http://campusmoodle.rgu.ac.uk/mod/url/view.php?id=3529827)

First, create a scatter map (the command works very similarly to scatter plots)

In [ ]:
# This should show a dot for every station in the dataset
fig = px.scatter_mapbox(canada,
                        lat="Latitude",
                        lon="Longitude")
fig.show()

Many of the same arguments from scatter plots are enabled on a scatter map, such as `size`, `size_max`, and `title`

In [ ]:
# Now the size of each dot changes depending on the amount of precipitation in 2017
fig = px.scatter_mapbox(canada[canada['Year'] == 2017],
                        lat="Latitude",
                        lon="Longitude",
                        size='Total Precipitation (mm)',
                        size_max=10,
                        title='Rainfall Over Canada (2017)')
fig.show()

Maps also have a `zoom` options which allows us to set the zoom level in advance

In [ ]:
# A zoom value of 2 in theory allows you to see "mainland" Canada full
# You may need to scroll up to see all stations
fig = px.scatter_mapbox(canada[canada['Year'] == 2017],
                        lat="Latitude",
                        lon="Longitude",
                        zoom=2,
                        size='Total Precipitation (mm)',
                        size_max=10,
                        title='Rainfall Over Canada (2017)')
fig.show()

We can add animations to maps in the same way as scatter plots

In [ ]:
fig = px.scatter_mapbox(canada,
                        lat="Latitude",
                        lon="Longitude",
                        zoom=2,
                        size='Total Precipitation (mm)',
                        size_max=10,
                        animation_frame='Year',
                        animation_group='Station Name',
                        title='Rainfall Over Canada')
fig.show()

Output hidden; open in https://colab.research.google.com to view.

In the 70's, I see that two dots appear in Europe!

- Can you figure out why by means of dataset manipulation?
- Is there any easy way to filter these out or to warn the user about this issue?

### Answer

The first thing we need to know is the latitude (min-max) and longitude (min-max) of Canada

According to [this site](https://www12.statcan.gc.ca/census-recensement/2011/ref/dict/geo016-eng.cfm), "For the land mass of Canada, latitudes range from about 42°N to 83°N and longitudes range from approximately 53°W to 141°W"

Therefore, any longitude value in the table that is **NOT** between -53 and -141 (west coordinates are represented in negative numbers, thanks UK!) will be in Europe!

In [ ]:
europe = canada[canada['Longitude']>-53]
europe

Station Name  Latitude  Longitude Province  \
2263             CAPE BROYLE    47.100    -52.933       NL   
2280                LOGY BAY    47.624    -52.664       NL   
2283   NORTH EAST POND RIVER    47.633    -52.833       NL   
2284           PETTY HARBOUR    47.467    -52.717       NL   
2285           PIERRES BROOK    47.283    -52.817       NL   
...                      ...       ...        ...      ...   
92847  ST JOHNS WEST CLIMATE    47.513    -52.783       NL   
93976      ST. JOHN'S INTL A    47.619    -52.753       NL   
93977  ST JOHNS WEST CLIMATE    47.513    -52.783       NL   
95078      ST. JOHN'S INTL A    47.619    -52.753       NL   
95079  ST JOHNS WEST CLIMATE    47.513    -52.783       NL   

       Mean Temperature (C)  Total Precipitation (mm)  Year  
2263                    NaN                      99.7  1971  
2280                   15.7                      37.1  1971  
2283                   15.3                     128.8  1971  
2284                    NaN                     139.9  1971  
2285                    NaN                     140.1  1971  
...                     ...                       ...   ...  
92847                  12.5                     111.9  2015  
93976                  15.5                     101.0  2016  
93977                  15.7                      70.7  2016  
95078                  16.3                     101.5  2017  
95079                  16.5                      99.0  2017  

[321 rows x 7 columns]

Maybe -53 was too strict, so let's look for -20!

In [ ]:
europe = canada[canada['Longitude']>-20]
europe

Station Name  Latitude  Longitude Province  \
2312         LAHR A GERMANY    48.367     -7.833      NaN   
2313   SOELLINGEN A GERMANY    48.767      8.100      NaN   
4641         LAHR A GERMANY    48.367     -7.833      NaN   
4642   SOELLINGEN A GERMANY    48.767      8.100      NaN   
7205         LAHR A GERMANY    48.367     -7.833      NaN   
7206   SOELLINGEN A GERMANY    48.767      8.100      NaN   
9812         LAHR A GERMANY    48.367     -7.833      NaN   
9813   SOELLINGEN A GERMANY    48.767      8.100      NaN   
12362        LAHR A GERMANY    48.367     -7.833      NaN   
12363  SOELLINGEN A GERMANY    48.767      8.100      NaN   
14856        LAHR A GERMANY    48.367     -7.833      NaN   
14857  SOELLINGEN A GERMANY    48.767      8.100      NaN   
17286        LAHR A GERMANY    48.367     -7.833      NaN   
17287  SOELLINGEN A GERMANY    48.767      8.100      NaN   
19701        LAHR A GERMANY    48.367     -7.833      NaN   
19702  SOELLINGEN A GERMANY    48.767      8.100      NaN   
22082        LAHR A GERMANY    48.367     -7.833      NaN   
22083  SOELLINGEN A GERMANY    48.767      8.100      NaN   
24460        LAHR A GERMANY    48.367     -7.833      NaN   
24461  SOELLINGEN A GERMANY    48.767      8.100      NaN   
26917        LAHR A GERMANY    48.367     -7.833      NaN   
26918  SOELLINGEN A GERMANY    48.767      8.100      NaN   
29301        LAHR A GERMANY    48.367     -7.833      NaN   
29302  SOELLINGEN A GERMANY    48.767      8.100      NaN   
31698        LAHR A GERMANY    48.367     -7.833      NaN   
31699  SOELLINGEN A GERMANY    48.767      8.100      NaN   
34121        LAHR A GERMANY    48.367     -7.833      NaN   
34122  SOELLINGEN A GERMANY    48.767      8.100      NaN   
36485        LAHR A GERMANY    48.367     -7.833      NaN   
36486  SOELLINGEN A GERMANY    48.767      8.100      NaN   
38844        LAHR A GERMANY    48.367     -7.833      NaN   
38845  SOELLINGEN A GERMANY    48.767      8.100      NaN   
41212        LAHR A GERMANY    48.367     -7.833      NaN   
41213  SOELLINGEN A GERMANY    48.767      8.100      NaN   
43601        LAHR A GERMANY    48.367     -7.833      NaN   
43602  SOELLINGEN A GERMANY    48.767      8.100      NaN   
46016        LAHR A GERMANY    48.367     -7.833      NaN   
46017  SOELLINGEN A GERMANY    48.767      8.100      NaN   
48434        LAHR A GERMANY    48.367     -7.833      NaN   
48435  SOELLINGEN A GERMANY    48.767      8.100      NaN   
50791        LAHR A GERMANY    48.367     -7.833      NaN   
50792  SOELLINGEN A GERMANY    48.767      8.100      NaN   

       Mean Temperature (C)  Total Precipitation (mm)  Year  
2312                   20.8                      21.1  1971  
2313                   20.6                      20.5  1971  
4641                   19.7                      44.4  1972  
4642                   19.6                      88.8  1972  
7205                   19.4                      53.8  1973  
7206                   19.1                      93.4  1973  
9812                   18.1                      78.0  1974  
9813                   18.0                     108.6  1974  
12362                  19.7                      48.3  1975  
12363                  19.9                      51.6  1975  
14856                  21.1                      95.8  1976  
14857                  21.8                      71.5  1976  
17286                  18.3                     106.7  1977  
17287                  18.4                      56.2  1977  
19701                  17.8                     113.6  1978  
19702                  17.6                     136.8  1978  
22082                  18.3                      48.5  1979  
22083                  18.4                      40.3  1979  
24460                  16.6                     127.2  1980  
24461                  16.5                     175.0  1980  
26917                  18.3                      68.6  1981  
26918                  18.2

So apparently the dataset has some info from German stations!

According to [Wikipedia](https://en.wikipedia.org/wiki/Canadian_Forces_Base_Lahr), there are a couple Canadian military bases in Söllingen and Lahr, in Germany

While the first city seems to be correctly plotted (border with France), the second one is plotted in the middle of the ocean!

By looking at the table above, we can see that this is due to Lahr being mapped in longitude -7.833, which is not correct as Germany is located on positive longitude values

Let's change this (on the original dataset)

In [ ]:
canada['Longitude'] = canada['Longitude'].replace(-7.833, 7.833)
canada

Station Name  Latitude  Longitude Province  \
0                   ALBERT HEAD    48.400   -123.483       BC   
1        BAMBERTON OCEAN CEMENT    48.583   -123.517       BC   
2                    BEAR CREEK    48.500   -124.000       BC   
3      CENTRAL SAANICH VEYANESS    48.583   -123.417       BC   
4      CENTRAL SAANICH ISL VIEW    48.571   -123.373       BC   
...                         ...       ...        ...      ...   
95090                   GOOSE A    53.317    -60.417       NL   
95091            HOPEDALE (AUT)    55.450    -60.217       NL   
95092          MARY'S HARBOUR A    52.303    -55.848       NL   
95093                      NAIN    56.550    -61.683       NL   
95094                  WABUSH A    52.923    -66.865       NL   

       Mean Temperature (C)  Total Precipitation (mm)  Year  
0                       NaN                       5.0  1971  
1                       NaN                       9.7  1971  
2                      15.4                      20.9  1971  
3                       NaN                       9.4  1971  
4                       NaN                      11.4  1971  
...                     ...                       ...   ...  
95090                  15.8                     109.0  2017  
95091                  11.6                      83.2  2017  
95092                  14.5                      56.9  2017  
95093                  10.6                      38.3  2017  
95094                  12.9                     129.0  2017  

[95095 rows x 7 columns]

We can check the German bases again...

In [ ]:
germany = canada[canada['Longitude']>-20]
germany

Station Name  Latitude  Longitude Province  \
2312         LAHR A GERMANY    48.367     -7.833      NaN   
2313   SOELLINGEN A GERMANY    48.767      8.100      NaN   
4641         LAHR A GERMANY    48.367     -7.833      NaN   
4642   SOELLINGEN A GERMANY    48.767      8.100      NaN   
7205         LAHR A GERMANY    48.367     -7.833      NaN   
7206   SOELLINGEN A GERMANY    48.767      8.100      NaN   
9812         LAHR A GERMANY    48.367     -7.833      NaN   
9813   SOELLINGEN A GERMANY    48.767      8.100      NaN   
12362        LAHR A GERMANY    48.367     -7.833      NaN   
12363  SOELLINGEN A GERMANY    48.767      8.100      NaN   
14856        LAHR A GERMANY    48.367     -7.833      NaN   
14857  SOELLINGEN A GERMANY    48.767      8.100      NaN   
17286        LAHR A GERMANY    48.367     -7.833      NaN   
17287  SOELLINGEN A GERMANY    48.767      8.100      NaN   
19701        LAHR A GERMANY    48.367     -7.833      NaN   
19702  SOELLINGEN A GERMANY    48.767      8.100      NaN   
22082        LAHR A GERMANY    48.367     -7.833      NaN   
22083  SOELLINGEN A GERMANY    48.767      8.100      NaN   
24460        LAHR A GERMANY    48.367     -7.833      NaN   
24461  SOELLINGEN A GERMANY    48.767      8.100      NaN   
26917        LAHR A GERMANY    48.367     -7.833      NaN   
26918  SOELLINGEN A GERMANY    48.767      8.100      NaN   
29301        LAHR A GERMANY    48.367     -7.833      NaN   
29302  SOELLINGEN A GERMANY    48.767      8.100      NaN   
31698        LAHR A GERMANY    48.367     -7.833      NaN   
31699  SOELLINGEN A GERMANY    48.767      8.100      NaN   
34121        LAHR A GERMANY    48.367     -7.833      NaN   
34122  SOELLINGEN A GERMANY    48.767      8.100      NaN   
36485        LAHR A GERMANY    48.367     -7.833      NaN   
36486  SOELLINGEN A GERMANY    48.767      8.100      NaN   
38844        LAHR A GERMANY    48.367     -7.833      NaN   
38845  SOELLINGEN A GERMANY    48.767      8.100      NaN   
41212        LAHR A GERMANY    48.367     -7.833      NaN   
41213  SOELLINGEN A GERMANY    48.767      8.100      NaN   
43601        LAHR A GERMANY    48.367     -7.833      NaN   
43602  SOELLINGEN A GERMANY    48.767      8.100      NaN   
46016        LAHR A GERMANY    48.367     -7.833      NaN   
46017  SOELLINGEN A GERMANY    48.767      8.100      NaN   
48434        LAHR A GERMANY    48.367     -7.833      NaN   
48435  SOELLINGEN A GERMANY    48.767      8.100      NaN   
50791        LAHR A GERMANY    48.367     -7.833      NaN   
50792  SOELLINGEN A GERMANY    48.767      8.100      NaN   

       Mean Temperature (C)  Total Precipitation (mm)  Year  
2312                   20.8                      21.1  1971  
2313                   20.6                      20.5  1971  
4641                   19.7                      44.4  1972  
4642                   19.6                      88.8  1972  
7205                   19.4                      53.8  1973  
7206                   19.1                      93.4  1973  
9812                   18.1                      78.0  1974  
9813                   18.0                     108.6  1974  
12362                  19.7                      48.3  1975  
12363                  19.9                      51.6  1975  
14856                  21.1                      95.8  1976  
14857                  21.8                      71.5  1976  
17286                  18.3                     106.7  1977  
17287                  18.4                      56.2  1977  
19701                  17.8                     113.6  1978  
19702                  17.6                     136.8  1978  
22082                  18.3                      48.5  1979  
22083                  18.4                      40.3  1979  
24460                  16.6                     127.2  1980  
24461                  16.5                     175.0  1980  
26917                  18.3                      68.6  1981  
26918                  18.2

The replace didn't work!

Let's check the value stored there...

In [ ]:
germany.at[50791,'Longitude']

-7.832999999999999

The problem is that the real value contained is -7.832999999999998, not -7.833!

We can either replace again using the actual value, or instead apply a function to the column to calculate the absolute value of any value between -20 and 0

In [ ]:
# creating and testing the function
def correct(value):
    if -20<value<0:
        return abs(value)
    else:
        return value
correct(-7.833)

7.833

In [ ]:
# Applying the function
canada['Longitude'] = canada['Longitude'].apply(correct)
canada[canada['Station Name']=='LAHR A GERMANY']

Station Name  Latitude  Longitude Province  Mean Temperature (C)  \
2312   LAHR A GERMANY    48.367      7.833      NaN                  20.8   
4641   LAHR A GERMANY    48.367      7.833      NaN                  19.7   
7205   LAHR A GERMANY    48.367      7.833      NaN                  19.4   
9812   LAHR A GERMANY    48.367      7.833      NaN                  18.1   
12362  LAHR A GERMANY    48.367      7.833      NaN                  19.7   
14856  LAHR A GERMANY    48.367      7.833      NaN                  21.1   
17286  LAHR A GERMANY    48.367      7.833      NaN                  18.3   
19701  LAHR A GERMANY    48.367      7.833      NaN                  17.8   
22082  LAHR A GERMANY    48.367      7.833      NaN                  18.3   
24460  LAHR A GERMANY    48.367      7.833      NaN                  16.6   
26917  LAHR A GERMANY    48.367      7.833      NaN                  18.3   
29301  LAHR A GERMANY    48.367      7.833      NaN                  20.5   
31698  LAHR A GERMANY    48.367      7.833      NaN                  22.6   
34121  LAHR A GERMANY    48.367      7.833      NaN                  18.5   
36485  LAHR A GERMANY    48.367      7.833      NaN                  20.0   
38844  LAHR A GERMANY    48.367      7.833      NaN                  19.0   
41212  LAHR A GERMANY    48.367      7.833      NaN                  19.3   
43601  LAHR A GERMANY    48.367      7.833      NaN                  19.0   
46016  LAHR A GERMANY    48.367      7.833      NaN                  20.1   
48434  LAHR A GERMANY    48.367      7.833      NaN                  18.5   
50791  LAHR A GERMANY    48.367      7.833      NaN                  21.2   

       Total Precipitation (mm)  Year  
2312                       21.1  1971  
4641                       44.4  1972  
7205                       53.8  1973  
9812                       78.0  1974  
12362                      48.3  1975  
14856                      95.8  1976  
17286                     106.7  1977  
19701                     113.6  1978  
22082                      48.5  1979  
24460                     127.2  1980  
26917                      68.6  1981  
29301                     115.3  1982  
31698                      33.1  1983  
34121                      87.1  1984  
36485                      52.2  1985  
38844                     101.0  1986  
41212                      99.9  1987  
43601                     101.1  1988  
46016                     129.8  1989  
48434                      38.1  1990  
50791                      80.6  1991

Finally, we can re-run the map

In [ ]:
fig = px.scatter_mapbox(canada,
                        lat="Latitude",
                        lon="Longitude",
                        zoom=2,
                        size='Total Precipitation (mm)',
                        size_max=10,
                        animation_frame='Year',
                        animation_group='Station Name',
                        title='Rainfall Over Canada')
fig.show()

Output hidden; open in https://colab.research.google.com to view.